### ISSUE TITLE
[tune] Pandas sorting error inPB2

### ISSUE THREAD

**Issue by ameroyer** [ray-project/ray#37638]
### What happened + What you expected to happen

Hello, 
I've ran into a pandas error when using the PB2 scheduler. The[ problematic line ](https://github.com/ray-project/ray/blob/master/python/ray/tune/schedulers/pb2.py#L196) is:

`df["y"] = df.groupby(["Trial"] + list(bounds.keys()))["Reward"].diff()`

The error is essentially failing in sorting the keys from the groupby  object (where `sort=True` is the default in pandas):

`'<' not supported between instances of 'Trial' and 'str'`

When printing the dataframe, there is in fact a row where the Trial is an actual object rather than a string representation like it is done in [save_trial_state](https://github.com/ray-project/ray/blob/master/python/ray/tune/schedulers/pb2.py#L447). 

My current guess is the issue may come from the [_explore](https://github.com/ray-project/ray/blob/master/python/ray/tune/schedulers/pb2.py#L245) function (lines 239-245); In some rare cases, a new datapoint is added  to the dataframe where the Trial object is used directly, rather than its string representation.

**Note:** The symptom is similar to this [older issue](https://github.com/ray-project/ray/issues/14187) but the cause seems different

### Versions / Dependencies

ray 2.5.0
pandas 1.5.2

### Reproduction script

So far I couldn't easily extract a minimal example that fails consistently, but the core experiment is a simple test of PB2, very similar to the documentation example, only optimizing two parameters (learning rate and weight decay). 
Also, the issue seems to trigger particularly often when `synch = True` but this might just be due to the dynamics of the search algorithm

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by CodingWookie** [comment]
Hello, did you resolve this ?

I am getting the same error.

I looked at the previous issue, and checked my PB2 hparams bounds convention [lower bound, upper bound]but all  looks correct, not sure where the issue is. 

This occurs after 2-5 iterations, after the exploit function but before this
``
2024-03-31 17:53:16,687 - GP - INFO - initializing Y
2024-03-31 17:53:16,687 - GP - INFO - initializing inference method
2024-03-31 17:53:16,687 - GP - INFO - adding kernel and likelihood as parameters
``


I also tried ray 2.9.3, 2.10.0 and 3.0 nightly... 
I also tried a fresh virtual env, reinstalling GPY from pip , and setup.py but no effect. Error persists.

I am running Ubuntu 22.04LTS and python 3.11.8 , and virtual env.
Here are my installed dependencies.
``
absl-py                       2.1.0
aiodns                        3.1.1
aiofiles                      23.2.1
aiohttp                       3.9.3
aiohttp-cors                  0.7.0
aiosignal                     1.3.1
alabaster                     0.7.16
alembic                       1.13.1
annotated-types               0.6.0
anyio                         4.3.0
arrow                         1.3.0
ast-comments                  1.2.1
attrs                         23.2.0
Babel                         2.14.0
beautifulsoup4                4.12.3
blosc2                        2.5.1
bottle                        0.12.25
bounded-pool-executor         0.0.3
Brotli                        1.1.0
cachetools                    5.3.3
catboost                      1.2.3
ccxt                          4.2.82
certifi                       2024.2.2
cffi                          1.16.0
charset-normalizer            3.3.2
click                         8.1.7
cloudpickle                   3.0.0
colorama                      0.4.6
colorful                      0.5.6
colorlog                      6.8.2
ConfigSpace                   0.7.1
contourpy                     1.2.0
cryptography                  42.0.5
cycler                        0.12.1
Cython                        3.0.10
dash                          2.16.1
dash-bootstrap-components     1.3.1
dash_colorscales              0.0.4
dash-core-components          2.0.0
dash_daq                      0.5.0
dash-html-components          2.0.0
dash-table                    5.0.0
datasieve                     0.1.7
decorator                     5.1.1
distlib                       0.3.8
docutils                      0.20.1
dtale                         3.11.0
et-xmlfile                    1.1.0
Farama-Notifications          0.0.4
fastapi                       0.110.0
filelock                      3.13.3
Flask                         2.2.5
Flask-Compress                1.14
flask-ngrok                   0.0.25
fonttools                     4.50.0
frozenlist                    1.4.1
fsspec                        2024.3.1
future                        1.0.0
google-api-core               2.18.0
google-auth                   2.29.0
googleapis-common-protos      1.63.0
GPy                           1.13.1
graphviz                      0.20.3
greenlet                      3.0.3
grpcio                        1.62.1
gym                           0.26.2
gym-notices                   0.0.8
gymnasium                     0.29.1
h11                           0.14.0
hdbscan                       0.8.33
httpcore                      1.0.5
httpx                         0.27.0
idna                          3.6
imagesize                     1.4.1
importlib_metadata            7.1.0
iniconfig                     2.0.0
itsdangerous                  2.1.2
janus                         1.0.0
Jinja2                        3.1.3
joblib                        1.3.2
jsonschema                    4.21.1
jsonschema-specifications     2023.12.1
kaleido                       0.2.1
kiwisolver                    1.4.5
lightgbm                      4.3.0
lz4                           4.3.3
Mako                          1.3.2
Markdown                      3.6
markdown-it-py                3.0.0
MarkupSafe                    2.1.5
matplotlib                    3.8.3
mdurl                         0.1.2
missingno                     0.5.2
more-itertools                10.2.0
mpmath                        1.3.0
msgpack                       1.0.8
multidict                     6.0.5
ncps                          0.0.7
ndindex                       1.8
nest-asyncio                  1.6.0
networkx                      3.2.1
numexpr                       2.9.0
numpy                         1.26.4
numpydoc                      1.7.0
nvidia-cublas-cu12            12.1.3.1
nvidia-cuda-cupti-cu12        12.1.105
nvidia-cuda-nvrtc-cu12        12.1.105
nvidia-cuda-runtime-cu12      12.1.105
nvidia-cudnn-cu12             8.9.2.26
nvidia-cufft-cu12             11.0.2.54
nvidia-curand-cu12            10.3.2.106
nvidia-cusolver-cu12          11.4.5.107
nvidia-cusparse-cu12          12.1.0.106
nvidia-nccl-cu12              2.19.3
nvidia-nvjitlink-cu12         12.4.99
nvidia-nvtx-cu12              12.1.105
opencensus                    0.11.4
opencensus-context            0.1.3
openpyxl                      3.1.2
optuna                        3.6.0
optuna-dashboard              0.15.1
orjson                        3.9.15
packaging                     24.0
pandas                        2.2.1
pandas_ta                     0.3.14b0
paramz                        0.9.6
patsy                         0.5.6
PDPbox                        0.3.0
pillow                        10.2.0
pip                           24.0
platformdirs                  4.2.0
plotly                        5.20.0
pluggy                        1.4.0
ppscore                       1.1.0
pqdm                          0.2.0
prometheus_client             0.20.0
prompt-toolkit                3.0.36
proto-plus                    1.23.0
protobuf                      4.25.3
psutil                        5.9.8
py-cpuinfo                    9.0.0
py_find_1st                   1.1.6
py-spy                        0.3.14
pyarrow                       15.0.2
pyasn1                        0.6.0
pyasn1_modules                0.4.0
pycares                       4.4.0
pycoingecko                   3.1.0
pycparser                     2.22
pydantic                      2.6.4
pydantic_core                 2.16.3
Pygments                      2.17.2
PyJWT                         2.8.0
pyparsing                     3.1.2
pytest                        8.1.1
python-dateutil               2.9.0.post0
python-rapidjson              1.16
python-telegram-bot           21.0.1
pytorch-tabnet                4.1.0
pytz                          2024.1
PyYAML                        6.0.1
questionary                   2.0.1
ray                           2.10.0
referencing                   0.34.0
requests                      2.31.0
retrying                      1.3.4
rich                          13.7.1
rpds-py                       0.18.0
rsa                           4.9
sb3-contrib                   2.2.1
sbx                           0.1.1
schedule                      1.2.1
scikit-learn                  1.4.1.post1
scipy                         1.11.4
sdnotify                      0.3.2
seaborn                       0.13.2
setuptools                    69.2.0
six                           1.16.0
smart-open                    7.0.4
sniffio                       1.3.1
snowballstemmer               2.2.0
soupsieve                     2.5
Sphinx                        7.2.6
sphinx-rtd-theme              2.0.0
sphinxcontrib-applehelp       1.0.8
sphinxcontrib-devhelp         1.0.6
sphinxcontrib-htmlhelp        2.0.5
sphinxcontrib-jquery          4.1
sphinxcontrib-jsmath          1.0.1
sphinxcontrib-qthelp          1.0.7
sphinxcontrib-serializinghtml 1.1.10
SQLAlchemy                    2.0.29
squarify                      0.4.3
stable-baselines3             2.2.1
starlette                     0.36.3
statsmodels                   0.14.1
strsimpy                      0.2.1
sympy                         1.12
TA-Lib                        0.4.28
tables                        3.9.1
tabulate                      0.9.0
technical                     1.4.3
tenacity                      8.2.3
tensorboard                   2.16.2
tensorboard-data-server       0.7.2
tensorboardX                  2.6.2.2
threadpoolctl                 3.4.0
torch                         2.2.1
tqdm                          4.66.2
triton                        2.2.0
types-python-dateutil         2.9.0.20240316
typing_extensions             4.10.0
tzdata                        2024.1
tzlocal                       5.2
urllib3                       2.2.1
uvicorn                       0.29.0
virtualenv                    20.25.1
wcwidth                       0.2.13
websockets                    12.0
Werkzeug                      3.0.1
wrapt                         1.16.0
xarray                        2024.3.0
xgboost                       2.0.3
xlrd                          2.0.1
yarl                          1.9.4
zipp                          3.18.1
``

**Comment by CodingWookie** [comment]
Hi team, can we get some love over here :) @ameroyer @anyscalesam @xwjiang2010

**Comment by ameroyer** [comment]
Hi @CodingWookie  
If you had the same error as I described in the first post, then yes I fixed it by overriding the PB2  class and making sure that whatever is added to the trials with "_explore" is a string representation rather than a trial

Specifically the fix looks like this; just use ``FixedPB2`` instead of ``PB2`` and that issue should disappear

```python
from ray.tune.schedulers.pb2 import PB2

class FixedPB2(PB2):

    def _get_new_config(self, *args, **kwargs):
        new_config, extra = super()._get_new_config(*args, **kwargs)
        # fix: make sure there are no Trial objects in data.Trial after _explore step
        self.data.Trial = self.data.Trial.astype("str")
        return new_config, extra

    
```

**Comment by CodingWookie** [comment]
@ameroyer  You are a gem. 

I can confirm changing line 242  in pb2 from this : 
`lst = [[trial, result[self._time_attr]] + values + [score]]`

To this 
`lst = [[str(old)] + [new_T] + values + [new_Reward]] `

 indeeed fixes the mentioned error!

I upgraded my PC and tried to scale up the agents, and now I can sucesfully do so 
Thank you !
